### PR TITLE
Small fixes towards ASR to LLVM implementation of COMMON block

### LINKED ISSUE #1684: [SciPy]: Assertion `Ty && "Invalid GetElementPtrInst indices for type!"' failed.
```fortran
program main
    integer :: a, b
    common / block_1 / a, b
    a = 2
end program
```

> ASR -> LLVM

```console
$ lfortran ./tests/common4.f90 --show-llvm
lfortran: /home/pranavchiku/conda_root/envs/lf/include/llvm/IR/Instructions.h:897: llvm::Type* llvm::checkGEPType(llvm::Type*): Assertion `Ty && "Invalid GetElementPtrInst indices for type!"' failed.
Traceback (most recent call last):
  Binary file "/home/pranavchiku/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 1892, in ??
    return emit_llvm(arg_file, lfortran_pass_manager,
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 807, in ??
    = fe.get_llvm(input, lm, pass_manager, diagnostics);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 291, in LCompilers::FortranEvaluator::get_llvm(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&)
    Result<std::unique_ptr<LLVMModule>> res = get_llvm2(code, lm, pass_manager, diagnostics);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 313, in LCompilers::FortranEvaluator::get_llvm2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&)
    diagnostics, lm.files.back().in_filename);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 346, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 8475, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    v.visit_asr((ASR::asr_t&)asr);
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4751, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_asr(LCompilers::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4727, in ??
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4752, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_unit(LCompilers::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4474, in ??
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 1370, in LCompilers::ASRToLLVMVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    visit_symbol(*item.second);
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4754, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/pranavchiku/lfortran/src/libasr/../libasr/asr.h", line 4482, in ??
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 2840, in LCompilers::ASRToLLVMVisitor::visit_Program(LCompilers::ASR::Program_t const&)
    declare_vars(x);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3437, in void LCompilers::ASRToLLVMVisitor::declare_vars<LCompilers::ASR::Program_t>(LCompilers::ASR::Program_t const&, bool)
    allocate_array_members_of_struct(ptr, v->m_type);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3263, in LCompilers::ASRToLLVMVisitor::allocate_array_members_of_struct(llvm::Value*, LCompilers::ASR::ttype_t*)
    llvm::Value* ptr_member = llvm_utils->create_gep(ptr, idx);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/llvm_utils.cpp", line 125, in LCompilers::LLVMUtils::create_gep(llvm::Value*, int)
    return LLVM::CreateGEP(*builder, ds, idx_vec);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/llvm_utils.cpp", line 27, in LCompilers::LLVM::CreateGEP(llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&, llvm::Value*, std::vector<llvm::Value*, std::allocator<llvm::Value*> >&)
    return builder.CreateGEP(t2, x, idx);
  File "/home/pranavchiku/conda_root/envs/lf/include/llvm/IR/IRBuilder.h", line 1773, in llvm::IRBuilderBase::CreateGEP(llvm::Type*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, llvm::Twine const&)
    return Insert(GetElementPtrInst::Create(Ty, Ptr, IdxList), Name);
  File "/home/pranavchiku/conda_root/envs/lf/include/llvm/IR/Instructions.h", line 943, in llvm::GetElementPtrInst::Create(llvm::Type*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, llvm::Twine const&, llvm::Instruction*)
    NameStr, InsertBefore);
  File "/home/pranavchiku/conda_root/envs/lf/include/llvm/IR/Instructions.h", line 1147, in llvm::GetElementPtrInst::GetElementPtrInst(llvm::Type*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::Twine const&, llvm::Instruction*)
    ResultElementType(getIndexedType(PointeeType, IdxList)) {
  File "/home/pranavchiku/conda_root/envs/lf/include/llvm/IR/Instructions.h", line 1074, in llvm::GetElementPtrInst::getGEPReturnType(llvm::Type*, llvm::Value*, llvm::ArrayRef<llvm::Value*>)
    Type *PtrTy = PointerType::get(checkGEPType(getIndexedType(ElTy, IdxList)),
  File "/home/pranavchiku/conda_root/envs/lf/include/llvm/IR/Instructions.h", line 897, in llvm::checkGEPType(llvm::Type*)
    assert(Ty && "Invalid GetElementPtrInst indices for type!");
  File "./assert/assert.c", line 101, in __GI___assert_fail()
  File "./assert/assert.c", line 92, in __assert_fail_base()
  File "./stdlib/abort.c", line 79, in __GI_abort()
  File "./signal/../sysdeps/posix/raise.c", line 26, in __GI_raise()
  File "./nptl/pthread_kill.c", line 43, in __pthread_kill_implementation()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __GI___sigaction()
Abort: Signal SIGABRT (abort) received

Aborted (core dumped)
```

### LINKED ISSUE #1703: [SciPy]: LCOMPILERS_ASSERT failed, COMMON ASR to LLVM
```fortran
program main
    real :: a, b
    common / block_1 / a, b
end program main
```

```console
%lfortran ./tests/temp.f90 --show-llvm
LCOMPILERS_ASSERT failed: /Users/pgoswami/Desktop/lfortran/src/libasr/pass/subroutine_from_function.cpp
function visit_TranslationUnit(), line number 45 at 
x.m_global_scope->get_symbol(item)
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1703. Fixes #1684.

### CODE REVIEW

File: integration_tests/common_05.f90 [hunk 1]
```diff
@@ -29,7 +25,7 @@ program main
         a = a - 1
     end do 
 
-    do a = 1, c(1), d
+    do a = 1, c(1), 1
         d = a + b
     end do
 end program
```
Comment (certik) [inline integration_tests/common_05.f90:31]: Initialize all variables and then add tests of the kind `if (d /= 10) error stop`.

**Review by certik — APPROVED** [review]

**Comment by Pranavchiku** [comment]
To compile COMMON blocks to ASR we need to fix #1723

**Comment by Pranavchiku** [comment]
I guess we can merge this and then with #1723 being fixed, we will be able to generate LLVM for common blocks.

**Review by certik — APPROVED** [review]
Looks good, thanks!

**Pull request merged**